src/Core/Impl/IO/DirectoryInfoProxy.cs

```diff
@@ -40,6 +43,31 @@ public sealed class DirectoryInfoProxy : IDirectoryInfo {
                 .EnumerateFileSystemInfos()
                 .Select(CreateFileSystemInfoProxy);
 
+        public IEnumerable<IFileSystemInfo> EnumerateFileSystemInfos(string[] includeFiles, string[] excludeFiles) {
+            Matcher matcher = new Matcher();
+            matcher.AddIncludePatterns(includeFiles.IsNullOrEmpty() ? new[] { "**/*" } : includeFiles);
+            matcher.AddExcludePatterns(excludeFiles ?? Enumerable.Empty<string>());
+            PatternMatchingResult matchResult = SafeExecuteMatcher(matcher);
+            return matchResult.Files.Select((filePatternMatch) => {
+                var fileSystemInfo = _directoryInfo.GetFileSystemInfos(filePatternMatch.Stem).First();
+                return CreateFileSystemInfoProxy(fileSystemInfo);
+            });
+        }
+
+        private PatternMatchingResult SafeExecuteMatcher(Matcher matcher) {
```

Comment:
This is a little goofy, relying on the loop exiting for the final result. I'd recommend something like:

```csharp
var directoryInfo = new DirectoryInfoWrapper(_directoryInfo);
for (var retries = 5; retries > 0; retries--) {
    try {
        return matcher.Execute(directoryInfo);
    } catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException) { }
}
return new PatternMatchingResult(Enumerable.Empty<FilePatternMatch>());
```

The docs states that `Execute` will always return a result (even when nothing is found), so returning should be safe. Otherwise, the loop ends and you return a default.